### PR TITLE
Flux from End-to-End

### PR DESCRIPTION
**Pull request by kingdonb**
This draft is intended to be written in the style of https://www.envoyproxy.io/docs/envoy/v1.19.1/intro/life_of_a_request based on https://github.com/fluxcd/website/issues/586 – an issue in our Flux Security Audit Follow-Up project.

I am interested in feedback about the structure, style, and goals of the document. Welcome to reply in this PR with feedback of any kind.

Also I am aware of the potential for overlap between this document and those that are produced in #595, #641, and #711. (Aim not to repeat ourselves in this document; if there is already a suitable doc available to describe the feature, defer to it.)

### CODE REVIEW

File: content/en/docs/guides/flux-e2e.md
Comment (dholbach) [inline content/en/docs/guides/flux-e2e.md:14]: https://fluxcd.io/docs/contributing/flux/#acceptance-policy maybe?

File: content/en/docs/guides/flux-e2e.md
Comment (dholbach) [inline content/en/docs/guides/flux-e2e.md:16]: https://fluxcd.io/docs/security/?

**Comment by kingdonb** [comment]
Thanks for the feedback so far, @dholbach! I'm working on the next revision which will include a completed glossary at the top, and more stories down below.

There should be more content to review today, and I'm hoping to wrap the content (excepting diagrams) before EOD tomorrow.

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [reply content/en/docs/guides/flux-e2e.md]: 👍

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [inline content/en/docs/guides/flux-e2e.md:10]: Awkward sentence

File: content/en/docs/guides/flux-e2e.md
Comment (stefanprodan) [inline content/en/docs/guides/flux-e2e.md:110]: This section is more suitable in the install docs, where we can explain what `flux install/bootstrap --network-policy` does.

File: content/en/docs/guides/flux-e2e.md
Comment (pjbgf) [inline content/en/docs/guides/flux-e2e.md:39]: Given that we also support `Bucket`, it may be worth using a more generic word instead of commit.

Comment (yebyen) [reply content/en/docs/guides/flux-e2e.md:39]: I wanted to use the word Commit here because it's in the glossary above.

A goal of this document is to keep as a good entrypoint for learning about Flux without expanding every detail of Flux to every section. I think it's OK to say "commit data" here and avoid mixing up metaphors too early. We can bring in the more general word when we introduce the bucket for source controller. 👍

Thanks @pjbgf !

(But there should be a section that mentions the source controller and buckets.)

File: content/en/docs/guides/flux-e2e.md
Comment (fsequeira1) [inline content/en/docs/guides/flux-e2e.md:233]:
```suggestion
Setup Notifications: https://fluxcd.io/docs/guides/notifications/#git-commit-status
```

File: content/en/docs/guides/flux-e2e.md
Comment (fsequeira1) [inline content/en/docs/guides/flux-e2e.md:230]:
```suggestion
GCSP work like other notification providers except that they target a specific commit with their event. If you use the git commit status provider integration for GitHub, GitLab, Bitbucket (or any supported git providers), then you will see success or failure reported on each commit for any alerts that are configured.
```

File: content/en/docs/guides/flux-e2e.md
Comment (fsequeira1) [inline content/en/docs/guides/flux-e2e.md:222]:
```suggestion
Notification Provides are for outbound notifications (the likes of slack, ms teams, discord, etc.) they’re driven by alerts (which are another CRD in Flux) alerts create notifications from events, and all of the flux reconcilers generate events.
```

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [reply content/en/docs/guides/flux-e2e.md:39]: There is now a reference to the `Bucket` kind under (4), (6). And a link out to the `BucketSpec` mentioned adjacent to where the `GitRepository` step is covered in detail under (7).

**Comment by kingdonb** [comment]
There are still three `TODO` items throughout the text, where there is known to be an issue to address and it's explained there.

I think those things don't block the review. And/or I don't want to expand the scope of this PR to any further than this single document if I can avoid it.

I would like to address all of the TODOs before this is merged, though some of them creep outside of this single document, and it is late here so this is something for tomorrow. (I'm sure there is likely to be other feedback that needs to be addressed before this can merge, which is the other reason it's marked as ready for review with these issues still remaining, so there will be plenty of opportunity to get around to all of that.)

I think any glaring structural issues have already been resolved, and the narrative is pretty solid. The numbers at the top should all line up with the numbers of the headings throughout... if we are good with this structure and don't need to break any other sub-component or forgotten ideas out, there are a total of 16 stops on the End-to-End tour of Flux.

👍 Looking forward to your feedback if you can take a look at this tomorrow! (Feedback is welcome from anyone out here of course, but I'm assigning @hiddeco as we discussed at the dev meeting, I need someone to review who is not a co-author.)

**Comment by kingdonb** [comment]
I have a doubt that the numbered headings are a good idea, as they will generate anchor links, which will not be stable if the numbers are ever re-arranged. Perhaps we can keep the numbered list at the top, but strip the numbers from the tour-de-flux headers below and place actual linking references from the descriptions above to the headers below?

(WDYT before I change anything else)

**Comment by kingdonb** [comment]
Oh! And I forgot to test out mermaid diagrams, and use the diagrams created in

* https://github.com/fluxcd/flux2/pull/2440

(TODO 5/5)

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [inline content/en/docs/guides/flux-e2e.md:241]: (this is TODO 4/5)

**Comment by kingdonb** [comment]
I got some feedback outside of the thread, I may be spending some time to try to cinch up the table of contents and tie it back to the rest of the doc today, (but as it's a holiday here today I might not get to it until tomorrow either...)

File: content/en/docs/guides/flux-e2e.md
Comment (hiddeco) [inline content/en/docs/guides/flux-e2e.md:268]: I would avoid using "part <num>" in the titels, as they add little (contextual) value.

File: content/en/docs/guides/flux-e2e.md
Comment (aryan9600) [inline content/en/docs/guides/flux-e2e.md:252]: Might be helpful to explain that users never explicitly create a `HelmChart` and how the definition of a `HelmRelease` is what the SC uses to create a `HelmChart`.

File: content/en/docs/guides/flux-e2e.md
Comment (aryan9600) [inline content/en/docs/guides/flux-e2e.md:67]: I think it'd be more intuitive, if we explain IRC before since it's functionality is a pre-requisite for the IAC.

File: content/en/docs/guides/flux-e2e.md
Comment (pjbgf) [inline content/en/docs/guides/flux-e2e.md:194]: Quite liked this, makes it easy to refer to the component's main features/responsibilities. :+1: 
Maybe worth having a shorter version here and a link to the up-to-date information (SC's readme?) that may change over time.

**Comment by kingdonb** [comment]
I still have some feedback to go over here, but many of the suggestions have already been taken into account, and the Table of Contents has been reworked so that it is in a more suitable form.

This should make it easier to provide more substantive feedback. I won't be able to address the rest of the feedback before about this time tomorrow, I will aim to have everything incorporated by Friday for another round of reviews.

Welcome to post additional feedback before then, and I will try to address it all at once 👍

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [inline content/en/docs/guides/flux-e2e.md:43]: Should there be more coverage of multi-tenancy in this End-to-End than a single mention in the brief description of the Kustomize Controller? (I think so)

**Comment by kingdonb** [comment]
I'm making another round of updates this morning.

I don't want to block merging the guide as long as we're happy with the structure. But since structure includes headings, and headings can be used as permalinks, I really want to be sure the structure is right before we merge it, so I'm not creating dead links in the future by removing headings that don't make sense in here anymore.

The first example of this conundrum is in the section on NetworkPolicy: we're agreed this doc is not the best place for the most detailed information on the current state of NetworkPolicy defaults in Flux, but I do not want to block the merging of this doc on improvements required in other docs (unless that's the consensus, and I don't think it is likely to be... I think we're aiming to merge this doc soon.)

Anyway, I'm still interested in feedback of course, any type of feedback, but if reviewers are hoping to help get this merged soon, the most important type of feedback right now is about the structure.

I'll be addressing all of the TODOs one way or another right now by fixing up the text and/or opening issues, and taking care of what I can from the other issues that remain unresolved from the thread above... I'll do so this morning, and back in a few hours to ring the bell when I think it's ready for final review.

File: content/en/docs/guides/flux-e2e.md
Comment (dholbach) [inline content/en/docs/guides/flux-e2e.md:10]: Maybe link to the instructions for a standard Flux installation? Is it just `/docs/get-started`?

File: content/en/docs/guides/flux-e2e.md
Comment (dholbach) [inline content/en/docs/guides/flux-e2e.md:51]:
```suggestion
Examples of some Helm Controller features that directly leverage upstream features of Helm today include [Helm Chart Hooks][], Helm Release Lifecycle events and the optional health checking that's performed by `helm --wait` to determine if a release is successful, Helm tests, rollbacks and uninstalls, and an implementation of Helm's [Post Rendering][] feature that allows safety and security while using the Kustomize post-renderer in Flux Continuous Delivery pipelines (that is, without requiring untrusted execution of any external scripts or binaries).
```

File: content/en/docs/guides/flux-e2e.md
Comment (dholbach) [inline content/en/docs/guides/flux-e2e.md:127]:
```suggestion
For more information, see: [`flux create`](/docs/cmd/flux_create/).
```

File: content/en/docs/guides/flux-e2e.md
Comment (dholbach) [inline content/en/docs/guides/flux-e2e.md:236]:
```suggestion
**Needs a link to FluxCD documentation for Server-Side Apply process!** _The only current on-point reference is the [Server-side reconciliation is coming](/blog/2021/09/server-side-reconciliation-is-coming/) blog post.  This would likely be considered a gap in FluxCD project docs._
```

**Comment by dholbach** [comment]
This is a well-written document and a great start. I'd vote for removing TODOs and merging soon and resolve additional requests (like integrating diagrams) to a follow-up issue.

As this explains many really concepts well, I wonder if we should at least refer to this from `/docs/concepts`. And maybe copy/move some of the introductory bits there and to the controller docs too. Happy for this to be a follow-up PR too.

Another question is if this best-placed in the `/docs/guides` section.

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [reply content/en/docs/guides/flux-e2e.md:10]: Added in efe433a 👍

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [reply content/en/docs/guides/flux-e2e.md:62]: 👍 Agreed, this is in 4c484a7

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [reply content/en/docs/guides/flux-e2e.md:234]: We can take things out in a subsequent PR, and move it elsewhere 👍

File: content/en/docs/guides/flux-e2e.md
Comment (kingdonb) [reply content/en/docs/guides/flux-e2e.md]: The diagrams are in here now, there are formatting issues but we can resolve them subsequently.

For some reason everything is `height="1024"` in the Chrome debugger/html source.

I'm not sure if that looks better on larger screens, but on my 13" MBP it's quite cramped and unnecessary whitespace on either side of the document. I imagine it does look better though, at 1080p or 4k resolutions and above. (So maybe better to leave it this way?)

**Comment by kingdonb** [comment]
I think this belongs right above Guides, now that there is text which clarifies this is an overview and not a full guide.

I'm reading this through once to be sure we haven't left any stage instructions where content should go (I'm sure there is at least one left), then I'll yank the TODOs from the text and give it an LGTM. 👍

**Comment by kingdonb** [comment]
That's it, fully proofed and LGTM 👍 rebasing and squashing a few commits once more for merge.

**Comment by kingdonb** [comment]
I believe the only remaining feedback that was unanswered is about moving the document out of `Guides`

Adding one more commit to address that now.

**Comment by kingdonb** [comment]
This is ready for final reviews now.

File: content/en/docs/flux-e2e.md
Comment (dholbach) [inline content/en/docs/flux-e2e.md:339]:
```suggestion
#### Diagram: Cluster sync from Git
```

File: content/en/docs/flux-e2e.md
Comment (dholbach) [inline content/en/docs/flux-e2e.md:487]:
```suggestion
#### Diagram: Helm release upgrade from Git
```

**Review by dholbach — APPROVED** [review]
Aside from some minor suggestions, this looks great.

Thank you so much for driving this effort - this will do a lot to make life for contributors and users easier!

**Review by pjbgf — COMMENTED** [review]
LGTM for me - once the recommended changes from Dan are done. :+1: 

Really great work putting this together @kingdonb!

**Review by hiddeco — APPROVED** [review]
Very pleased with how the original lengthy document got shaped into one of the right proportions. Assuming @dholbach his nitpicks are addressed, this looks good to me. Great job @kingdonb, and all others that helped :bow: :1st_place_medal:

**Pull request merged**
